File: app/src/main/java/com/example/furnitures/calculator/container/list/ListTrickFragment.java
package com.example.furnitures.calculator.container.list;

import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.example.furnitures.R;
import com.example.furnitures.calculator.container.selection.ListTrickAdapter;
import com.example.furnitures.calculator.helper.SimpleItemTouchHelperCallback;
import com.example.furnitures.calculator.trick.FurnitureViewState;

/**
 * Copyright (c) 2017 fluidmobile GmbH. All rights reserved.
 */
public class ListTrickFragment
    extends Fragment {

    private RecyclerView recyclerView;
    private ListTrickContract.ViewModel viewModel;

    private SimpleItemTouchHelperCallback simpleItemTouchHelperCallback;
    private ItemTouchHelper itemTouchHelper;

    @NonNull
    public static ListTrickFragment newInstance() {
        return new ListTrickFragment();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        viewModel = ViewModelProviders.of(getActivity()).get(ListTrickViewModel.class);

        simpleItemTouchHelperCallback = new SimpleItemTouchHelperCallback(this, viewModel);
        itemTouchHelper = new ItemTouchHelper(simpleItemTouchHelperCallback);

        // Use ItemTouchHelper (convenience class / bequemlichkeits Klasse)
        // Class that make our RecyclerView swipable
        // Pass SimpleCallBack to create the two Methods
        // SimpleCallback wants two parameters -> drag and drop direction and swipe direction
    }

    @Nullable
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_select_tricks, container, false);
        recyclerView = view.findViewById(R.id.fragment_select_tricks__recycler);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ListTrickAdapter adapter = new ListTrickAdapter();
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));
        recyclerView.setAdapter(adapter);

        // Observed alle veränderungen der Liste und gibt sie an den Adapter weiter
        // submitList löst automatisch notifyDataSetChanged aus
        viewModel.getSelectedItemsViewState().observe(getViewLifecycleOwner(), newData -> {
            if (newData != null) adapter.submitList(newData);
        });

        itemTouchHelper.attachToRecyclerView(recyclerView);
    }

    public void onItemMoved(Integer fromPosition, Integer toPosition){
        viewModel.onItemMove(fromPosition, toPosition);
    }

    public void onItemDelete(FurnitureViewState furnitureViewState){
        viewModel.onItemDelete(furnitureViewState);
    }

    public FurnitureViewState getFurnitureAt(Integer position){
        return viewModel.getFurnitureAt(position);
    }


}

File: app/src/main/java/com/example/furnitures/calculator/bottombar/list/ListTrickFragment.kt
package com.example.furnitures.calculator.bottombar.list

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.furnitures.R
import com.example.furnitures.calculator.bottombar.selection.ListTrickAdapter
import com.example.furnitures.calculator.trick.FurnitureViewState

class ListTrickFragment : Fragment() {

    private lateinit var recyclerView: RecyclerView
    private lateinit var viewModel: ListTrickContract.ViewModel

    private lateinit var listTrickItemMoveCallback: ListTrickItemMoveCallback
    private lateinit var itemTouchHelper: ItemTouchHelper

    private val adapter = ListTrickAdapter(
        object : ListTrickAdapter.FurnitureModificationListener {
            override fun onItemRemove(furnitureItem: FurnitureViewState) {
                viewModel.removeFurnitureItem(furnitureItem)
            }

            override fun onItemMove(fromPosition: Int, toPosition: Int) {
                viewModel.changeFurnitureItemPosition(fromPosition, toPosition)
            }
        }
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProviders.of(activity!!).get(ListTrickViewModel::class.java)
        listTrickItemMoveCallback = ListTrickItemMoveCallback(adapter)
        itemTouchHelper = ItemTouchHelper(listTrickItemMoveCallback)

        // Use ItemTouchHelper (convenience class / bequemlichkeits Klasse)
        // Class that make our RecyclerView swipable
        // Pass SimpleCallBack to create the two Methods
        // SimpleCallback wants two parameters -> drag and drop direction and swipe direction
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_select_tricks, container, false)
        recyclerView = view.findViewById(R.id.fragment_select_tricks__recycler)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        recyclerView.setHasFixedSize(true)
        recyclerView.layoutManager = LinearLayoutManager(requireContext())
        recyclerView.adapter = adapter

        // Observed alle veränderungen der Liste und gibt sie an den Adapter weiter
        // submitList löst automatisch notifyDataSetChanged aus
        viewModel.getSelectedItemsViewState().observe(viewLifecycleOwner, Observer{ newData
            -> if (newData != null) adapter.onFurnitureItemsUpdate(newData)
        })

        itemTouchHelper.attachToRecyclerView(recyclerView)
    }

    companion object {
        fun newInstance(): ListTrickFragment {
            return ListTrickFragment()
        }
    }


}
